modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/odata-util.es.js

```diff
@@ -0,0 +1,273 @@
+						{
+							lastNodeWasGroup: true,
+							prevConjunction: queryAST.type,
+							queryAST: queryAST.value.right
+						}
+					)
+				]
+			}
+		];
+	}
+	else {
+		const childType = getChildNodeTypeName(queryAST);
+
+		returnValue = [
+			{
+				conjunctionName: CONJUNCTIONS.includes(childType) ?
+					childType :
+					AND,
+				items: [
+					...toCriteriaMap(
+						{
+							lastNodeWasGroup: true,
+							prevConjunction,
+							queryAST: queryAST.value
+						}
+					)
+				]
+			}
+		];
+	}
+
+	return returnValue;
+};
+
+function isCriteriaMapGroup(criteriaMapArray) {
+	return !!criteriaMapArray[0].items;
+}
+
+function isNotConjunction(nodeType) {
+	return !CONJUNCTIONS.includes(getTypeName(nodeType));
+}
+
+const getNextNonGroupNodeType = queryAST => {
+	let returnValue;
+
+	if (queryAST.value.type === 'BoolParentExpression') {
+		returnValue = getNextNonGroupNodeType(queryAST.value);
+	}
+	else {
+		returnValue = queryAST.value.left ?
+			queryAST.value.left.type :
+			queryAST.value.type;
+	}
+
+	return returnValue;
+};
+
+const getChildNodeTypeName = query =>
+	oDataTransformationMap[query.value.type].name;
+
+const getTypeName = type => oDataTransformationMap[type].name;
+
+const operatorTransformation = ({queryAST}) => {
+	return [
+		{
+			operatorName: oDataTransformationMap[queryAST.type].name,
+			propertyName: queryAST.value.left.raw,
+			value: queryAST.value.right.raw.replace(/['"]+/g, '')
+		}
+	];
+};
+
+const skipGroup = (queryAST, prevConjunction) => ({
+	lastNodeWasGroup: true,
+	prevConjunction,
+	queryAST: queryAST.value
+});
+
+const toCriteriaMap = ({
+	lastNodeWasGroup = false,
+	prevConjunction,
+	queryAST
+}) => {
+	const oDataParserType = oDataTransformationMap[queryAST.type];
+
+	return oDataParserType.transformationFunction(
+		{
+			lastNodeWasGroup,
+			prevConjunction,
+			queryAST
+		}
+	);
+};
+
+const wrapInCriteriaMapGroup = criteriaMapArray => (
+	{
+		conjunctionName: AND,
+		items: criteriaMapArray
+	}
+);
+
+const oDataTransformationMap = {
+	AndExpression: {
+		name: AND,
+		transformationFunction: comparatorTransformation
+	},
+	BoolParenExpression: {
+		name: GROUP,
+		transformationFunction: groupTransformation
+	},
+	EqualsExpression: {
+		name: EQ,
+		transformationFunction: operatorTransformation
+	},
+	GreaterOrEqualsExpression: {
+		name: GE,
+		transformationFunction: operatorTransformation
+	},
+	GreaterThanExpression: {
+		name: GT,
+		transformationFunction: operatorTransformation
+	},
+	LesserOrEqualsExpression: {
+		name: LE,
+		transformationFunction: operatorTransformation
+	},
+	LesserThanExpression: {
+		name: LT,
+		transformationFunction: operatorTransformation
+	},
+	NotEqualsExpression: {
+		name: NE,
+		transformationFunction: operatorTransformation
+	},
+	OrExpression: {
+		name: OR,
+		transformationFunction: comparatorTransformation
+	}
+};
+
+export const buildQueryString = (queryItems, queryConjunction) => {
+	let queryString = '';
+
+	queryItems.forEach(
```

Comment:
could we use a reduce function here? Might make it a little more straight forward.

Reply:
So since check source formatter only allows a single return per function, using reduce is basically identical. Unless you had something different in mind? But using reduce in this context does make more sense.

```javascript
	return queryItems.reduce(
		(queryString, item, index) => {
			const {
				conjunctionName,
				items,
				operatorName,
				propertyName,
				value
			} = item;

			if (index > 0) {
				queryString = queryString.concat(` ${queryConjunction} `);
			}

			if (conjunctionName) {
				queryString = queryString.concat(
					`(${buildQueryString(items, conjunctionName)})`
				);
			}
			else if (RELATIONAL_OPERATORS.includes(operatorName)) {
				queryString = queryString.concat(
					`${propertyName} ${operatorName} '${value}'`
				);
			}
			else if (FUNCTIONAL_OPERATORS.includes(operatorName)) {
				queryString = queryString.concat(
					`${operatorName} (${propertyName}, '${value}')`
				);
			}

			return queryString;
		},
		''
	);
```